### PR TITLE
Jvenstad/application locking revised

### PR DESCRIPTION
**Pull request by jonmv**
@bratseth I ask for nothing but your careful consideration. 

**Pros:**
 * Wrap and hide `Lock`; less room for error, less code.
 * Hide `LockedApplication`; make it more clear that these are not for keeping.
 * Makes it easier to refactor later, since things are more controlled. 

*Cons:*
 * Functional style way of declaring `lockedApplication` variables (as lambda parameter)?
 * Can't return an `Application` while holding its `Lock` — not really a problem?

### CODE REVIEW

**Comment by jonmv** [comment]
@mpolden do you have any opinions or comments?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java [hunk 1]
```diff
@@ -288,9 +268,10 @@ public Application createApplication(ApplicationId id, Optional<NToken> token) {
     public ActivateResult deployApplication(ApplicationId applicationId, Zone zone,
                                             ApplicationPackage applicationPackage, DeployOptions options) {
         try (Lock lock = lock(applicationId)) {
-            LockedApplication application = get(applicationId, lock).orElse(new LockedApplication(
+            // TODO: Shouldn't this go through the above method? Seems you can cheat the checks here ... ?
+            LockedApplication application = get(applicationId).map(application1 -> new LockedApplication(application1, lock)).orElse(new LockedApplication(
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: Will we break anything if this is tightened?

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java [hunk 2]
```diff
@@ -536,6 +517,30 @@ public void store(LockedApplication application) {
...
+     * Acquire a locked application to modify and store, if there is an application with the given id.
+     *
+     * @param applicationId Id of the application to lock and get.
+     * @param actions Things to do with the locked application.
+     */
+    public void lockedIfPresent(ApplicationId applicationId, Consumer<LockedApplication> actions) {
+        try (Lock lock = lock(applicationId)) {
+            get(applicationId).map(application -> new LockedApplication(application, lock)).ifPresent(actions);
+        }
+    }
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: New way to get LockedApplication.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java [hunk 3]
```diff
@@ -536,6 +517,30 @@ public void store(LockedApplication application) {
...
+     *
+     * @param applicationId Id of the application to lock and require.
+     * @param actions Things to do with the locked application.
+     */
+    public void lockedOrThrow(ApplicationId applicationId, Consumer<LockedApplication> actions) {
+        try (Lock lock = lock(applicationId)) {
+            actions.accept(new LockedApplication(require(applicationId), lock));
+        }
+    }
+
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: Same as above.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 4]
```diff
@@ -180,4 +141,47 @@ public Version deployVersionFor(DeploymentJobs.JobType jobType, Controller contr
...
 
+    /** Don't expose non-leaf sub-objects. */
+    private LockedApplication with(Deployment deployment) {
+        Map<Zone, Deployment> deployments = new LinkedHashMap<>(deployments());
+        deployments.put(deployment.zone(), deployment);
+        return new LockedApplication(new Builder(this).with(deployments));
+    }
+
+
+    private static class Builder {
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: Not sure this was a real improvement :p

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 5]
```diff
@@ -78,8 +79,7 @@ public DeploymentTrigger(Controller controller, CuratorDb curator, Clock clock)
      * @param report information about the job that just completed
      */
     public void triggerFromCompletion(JobReport report) {
-        try (Lock lock = applications().lock(report.applicationId())) {
-            LockedApplication application = applications().require(report.applicationId(), lock);
+        applications().lockedOrThrow(report.applicationId(), application -> {
             application = application.withJobCompletion(report, clock.instant(), controller);
```
Comment (jonmv) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: Typical example of changed usage.

**Comment by mpolden** [comment]
> @mpolden do you have any opinions or comments?

Only that I don't see the point of these countless cleanups. I think refactoring should happen in the context of working on something worthwhile, like a new feature or fixing a bug, while this (and many other PRs) appear to change code for the sake of changing it. Just my two cents.

Code is infinitely malleable and will never be perfect. :-)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java [hunk 6]
```diff
@@ -17,79 +19,55 @@
...
 import java.util.Map;
-import java.util.Objects;
 import java.util.Optional;
 
 /**
  * A combination of an application instance and a lock for that application. Provides methods for updating application
  * fields.
  *
- * @author mpolden
+ * @author mpolden, jvenstad
```
Comment (bratseth) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/LockedApplication.java]: Fyi, one @author line per author

**Review by bratseth — APPROVED** [review]
Ok; fine  :-)
I think the diff shows quite clearly that this is an improvement ...

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java (hunk 1 above)
Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: Any comment on this, though?

Comment (jonmv) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: It seems you can create an application with a non-existent tenant here, by shortcutting createApplication. 
This method is called with an application id with a tenant name parsed directly from the URI. 
You won't be able to deploy this to prod, but you'll get an invalid application entry, and one which you can deploy to dev, etc..

Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/ApplicationController.java]: Ok, there's probably no particular good reason it's like this ...

**Pull request merged**